File: src/lib/predefinedTests.ts
// Predefined test templates for data quality validation
export interface TestTemplate {
    id: string;
    name: string;
    category: 'completeness' | 'integrity' | 'quality' | 'statistical' | 'business';
    severity: 'HIGH' | 'MEDIUM' | 'LOW';
    description: string;
    isGlobal: boolean;
    generateSQL: (config: any) => string | null;
}

export const PREDEFINED_TESTS: Record<string, TestTemplate> = {
    row_count_match: {
        id: 'row_count_match',
        name: 'Row Count Match',
        category: 'completeness',
        severity: 'HIGH',
        description: 'Verify source and target row counts match',
        isGlobal: true,
        generateSQL: () => null // Handled programmatically
    },

    no_nulls_required: {
        id: 'no_nulls_required',
        name: 'No NULLs in Required Fields',
        category: 'completeness',
        severity: 'HIGH',
        description: 'Check required columns have no NULL values',
        isGlobal: true,
        generateSQL: (config) => {
            if (!config.required_columns || config.required_columns.length === 0) {
                return null;
            }
            const conditions = config.required_columns
                .map((col: string) => `${col} IS NULL`)
                .join(' OR ');
            return `
        SELECT * FROM \`${config.fullTableName}\`
        WHERE ${conditions}
        LIMIT 100
      `;
        }
    },

    no_duplicates_pk: {
        id: 'no_duplicates_pk',
        name: 'No Duplicate Primary Keys',
        category: 'integrity',
        severity: 'HIGH',
        description: 'Ensure primary key uniqueness',
        isGlobal: true,
        generateSQL: (config) => {
            if (!config.primary_key_columns || config.primary_key_columns.length === 0) {
                return null;
            }
            const pkCols = config.primary_key_columns.join(', ');
            return `
        SELECT ${pkCols}, COUNT(*) as duplicate_count
        FROM \`${config.fullTableName}\`
        GROUP BY ${pkCols}
        HAVING COUNT(*) > 1
      `;
        }
    },

    referential_integrity: {
        id: 'referential_integrity',
        name: 'Referential Integrity',
        category: 'integrity',
        severity: 'HIGH',
        description: 'Validate foreign key relationships',
        isGlobal: false,
        generateSQL: (config) => {
            if (!config.foreign_key_checks || Object.keys(config.foreign_key_checks).length === 0) {
                return null;
            }

            const checks = Object.entries(config.foreign_key_checks).map(([fkCol, ref]: [string, any]) => `
        SELECT 
          '${fkCol}' as fk_column, 
          ${fkCol} as invalid_value,
          COUNT(*) as occurrence_count
        FROM \`${config.fullTableName}\` t
        WHERE ${fkCol} IS NOT NULL
        AND NOT EXISTS (
          SELECT 1 FROM \`${ref.table}\` r
          WHERE r.${ref.column} = t.${fkCol}
        )
        GROUP BY ${fkCol}
      `).join(' UNION ALL ');

            return checks;
        }
    },

    numeric_range: {
        id: 'numeric_range',
        name: 'Numeric Range Validation',
        category: 'quality',
        severity: 'MEDIUM',
        description: 'Check numeric values are within expected ranges',
        isGlobal: false,
        generateSQL: (config) => {
            if (!config.numeric_range_checks || Object.keys(config.numeric_range_checks).length === 0) {
                return null;
            }

            const conditions = Object.entries(config.numeric_range_checks)
                .map(([col, range]: [string, any]) =>
                    `(${col} < ${range.min} OR ${col} > ${range.max})`
                )
                .join(' OR ');

            return `
        SELECT * FROM \`${config.fullTableName}\`
        WHERE ${conditions}
        LIMIT 100
      `;
        }
    },

    date_range: {
        id: 'date_range',
        name: 'Date Range Validation',
        category: 'quality',
        severity: 'MEDIUM',
        description: 'Validate dates are within expected range',
        isGlobal: false,
        generateSQL: (config) => {
            if (!config.date_range_checks || Object.keys(config.date_range_checks).length === 0) {
                return null;
            }

            const conditions = Object.entries(config.date_range_checks)
                .map(([col, range]: [string, any]) =>
                    `(${col} < '${range.min_date}' OR ${col} > '${range.max_date}')`
                )
                .join(' OR ');

            return `
        SELECT * FROM \`${config.fullTableName}\`
        WHERE ${conditions}
        LIMIT 100
      `;
        }
    },

    pattern_validation: {
        id: 'pattern_validation',
        name: 'Pattern Validation',
        category: 'quality',
        severity: 'MEDIUM',
        description: 'Check string patterns (email, phone, etc.)',
        isGlobal: false,
        generateSQL: (config) => {
            if (!config.pattern_checks || Object.keys(config.pattern_checks).length === 0) {
                return null;
            }

            const conditions = Object.entries(config.pattern_checks)
                .map(([col, pattern]: [string, any]) =>
                    `NOT REGEXP_CONTAINS(CAST(${col} AS STRING), r'${pattern}')`
                )
                .join(' OR ');

            return `
        SELECT * FROM \`${config.fullTableName}\`
        WHERE ${conditions}
        LIMIT 100
      `;
        }
    },

    outlier_detection: {
        id: 'outlier_detection',
        name: 'Statistical Outlier Detection',
        category: 'statistical',
        severity: 'LOW',
        description: 'Detect statistical outliers using standard deviation',
        isGlobal: false,
        generateSQL: (config) => {
            if (!config.outlier_columns || config.outlier_columns.length === 0) {
                return null;
            }

            const col = config.outlier_columns[0]; // Check first column
            return `
        WITH stats AS (
          SELECT 
            AVG(${col}) as mean,
            STDDEV(${col}) as stddev
          FROM \`${config.fullTableName}\`
          WHERE ${col} IS NOT NULL
        )
        SELECT t.* 
        FROM \`${config.fullTableName}\` t, stats
        WHERE ABS(t.${col} - stats.mean) > 3 * stats.stddev
        LIMIT 100
      `;
        }
    }
};

// Helper to get enabled tests for a mapping
export function getEnabledTests(enabledTestIds?: string[]): TestTemplate[] {
    if (!enabledTestIds || enabledTestIds.length === 0) {
        // Return all global tests by default
        return Object.values(PREDEFINED_TESTS).filter(t => t.isGlobal);
    }

    return enabledTestIds
        .map(id => PREDEFINED_TESTS[id])
        .filter(t => t !== undefined);
}

File: backend/app/tests/predefined_tests.py
"""Predefined test templates for data quality validation."""
from typing import Dict, List, Callable, Optional


class TestTemplate:
    """Template for a predefined test."""
    
    def __init__(
        self,
        test_id: str,
        name: str,
        category: str,
        severity: str,
        description: str,
        is_global: bool,
        generate_sql: Callable[[Dict], Optional[str]]
    ):
        self.id = test_id
        self.name = name
        self.category = category
        self.severity = severity
        self.description = description
        self.is_global = is_global
        self.generate_sql = generate_sql


# Predefined test templates
PREDEFINED_TESTS = {
    'row_count_match': TestTemplate(
        test_id='row_count_match',
        name='Row Count Match',
        category='completeness',
        severity='HIGH',
        description='Verify source and target row counts match',
        is_global=True,
        generate_sql=lambda config: None  # Handled programmatically
    ),
    
    'no_nulls_required': TestTemplate(
        test_id='no_nulls_required',
        name='No NULLs in Required Fields',
        category='completeness',
        severity='HIGH',
        description='Check required columns have no NULL values',
        is_global=True,
        generate_sql=lambda config: (
            f"""
            SELECT * FROM `{config['full_table_name']}`
            WHERE {' OR '.join(f"{col} IS NULL" for col in config.get('required_columns', []))}
            LIMIT 100
            """ if config.get('required_columns') else None
        )
    ),
    
    'no_duplicates_pk': TestTemplate(
        test_id='no_duplicates_pk',
        name='No Duplicate Primary Keys',
        category='integrity',
        severity='HIGH',
        description='Ensure primary key uniqueness',
        is_global=True,
        generate_sql=lambda config: (
            f"""
            SELECT {', '.join(config['primary_key_columns'])}, COUNT(*) as duplicate_count
            FROM `{config['full_table_name']}`
            GROUP BY {', '.join(config['primary_key_columns'])}
            HAVING COUNT(*) > 1
            """ if config.get('primary_key_columns') else None
        )
    ),
    
    'table_exists': TestTemplate(
        test_id='table_exists',
        name='Table exists (smoke)',
        category='smoke',
        severity='HIGH',
        description='Verify the target table exists and is accessible',
        is_global=False,
        generate_sql=lambda config: f"SELECT TRUE FROM `{config['full_table_name']}` LIMIT 1"
    ),

    # --- SCD1 Tests ---
    'scd1_primary_key_null': TestTemplate(
        test_id='scd1_primary_key_null',
        name='Primary Key NOT NULL',
        category='completeness',
        severity='HIGH',
        description='Check composite primary key for NULL values',
        is_global=False,
        generate_sql=lambda config: (
            f"""
            SELECT * FROM `{config['full_table_name']}`
            WHERE ({' || '.join([f"SAFE_CAST({col} AS STRING)" for col in config['primary_keys']])}) IS NULL
            LIMIT 100
            """ if config.get('primary_keys') else None
        )
    ),

    'scd1_primary_key_unique': TestTemplate(
        test_id='scd1_primary_key_unique',
        name='Primary Key uniqueness',
        category='integrity',
        severity='HIGH',
        description='Ensure composite primary key uniqueness',
        is_global=False,
        generate_sql=lambda config: (
            f"""
            SELECT 
                ({' || '.join([f"IFNULL(SAFE_CAST({col} AS STRING), '')" for col in config['primary_keys']])}) as primary_key,
                COUNT(*) as duplicate_count
            FROM `{config['full_table_name']}`
            GROUP BY 1
            HAVING COUNT(*) > 1
            LIMIT 100
            """ if config.get('primary_keys') else None
        )
    ),

    # --- SCD2 Tests (11 + 4 Structural) ---
    'scd2_primary_key_null': TestTemplate(
        test_id='scd2_primary_key_null',
        name='Primary Key NOT NULL',
        category='completeness',
        severity='HIGH',
        description='Check SCD2 primary key for NULL values',
        is_global=False,
        generate_sql=lambda config: (
            f"""
            SELECT * FROM `{config['full_table_name']}`
            WHERE ({' || '.join([f"SAFE_CAST({col} AS STRING)" for col in config['primary_keys']])}) IS NULL
            LIMIT 100
            """ if config.get('primary_keys') else None
        )
    ),

    'scd2_begin_date_null': TestTemplate(
        test_id='scd2_begin_date_null',
        name='Begin effective datetime NOT NULL',
        category='completeness',
        severity='HIGH',
        description='Check SCD2 begin date for NULL values',
        is_global=False,
        generate_sql=lambda config: f"SELECT * FROM `{config['full_table_name']}` WHERE {config['begin_date_column']} IS NULL LIMIT 100"
    ),

    'scd2_end_date_null': TestTemplate(
        test_id='scd2_end_date_null',
        name='End effective datetime NOT NULL',
        category='completeness',
        severity='HIGH',
        description='Check SCD2 end date for NULL values',
        is_global=False,
        generate_sql=lambda config: f"SELECT * FROM `{config['full_table_name']}` WHERE {config['end_date_column']} IS NULL LIMIT 100"
    ),

    'scd2_flag_null': TestTemplate(
        test_id='scd2_flag_null',
        name='Current row flag NOT NULL',
        category='completeness',
        severity='HIGH',
        description='Check SCD2 active flag for NULL values',
        is_global=False,
        generate_sql=lambda config: f"SELECT * FROM `{config['full_table_name']}` WHERE {config['active_flag_column']} IS NULL LIMIT 100"
    ),

    'scd2_one_current_row': TestTemplate(
        test_id='scd2_one_current_row',
        name='One current row per Primary Key',
        category='integrity',
        severity='HIGH',
        description='Ensure exactly one active record per primary key',
        is_global=False,
        generate_sql=lambda config: (
            f"""
            SELECT {' , '.join(config['primary_keys'])}, COUNTIF(SAFE_CAST({config['active_flag_column']} AS STRING) IN ('true', 'TRUE', 'Y', '1')) as active_count
            FROM `{config['full_table_name']}`
            GROUP BY {' , '.join(config['primary_keys'])}
            HAVING active_count <> 1
            LIMIT 100
            """
        )
    ),

    'scd2_current_date_check': TestTemplate(
        test_id='scd2_current_date_check',
        name='Current rows end on 2099-12-31',
        category='validity',
        severity='HIGH',
        description='Ensure active rows have the high-watermark end date',
        is_global=False,
        generate_sql=lambda config: (
            f"""
            SELECT * FROM `{config['full_table_name']}`
            WHERE SAFE_CAST({config['active_flag_column']} AS STRING) IN ('true', 'TRUE', 'Y', '1')
            AND CAST({config['end_date_column']} AS STRING) NOT LIKE '2099-12-31%'
            LIMIT 100
            """
        )
    ),

    'scd2_invalid_flag_combination': TestTemplate(
        test_id='scd2_invalid_flag_combination',
        name='No invalid current-row combinations',
        category='validity',
        severity='HIGH',
        description='Ensure active flag is only set for high-watermark end dates',
        is_global=False,
        generate_sql=lambda config: (
            f"""
            SELECT * FROM `{config['full_table_name']}`
            WHERE SAFE_CAST({config['active_flag_column']} AS STRING) IN ('true', 'TRUE', 'Y', '1')
            AND CAST({config['end_date_column']} AS STRING) NOT LIKE '2099-12-31%'
            LIMIT 100
            """
        )
    ),

    'scd2_date_order': TestTemplate(
        test_id='scd2_date_order',
        name='Begin < End datetime',
        category='validity',
        severity='HIGH',
        description='Ensure BeginEffDateTime < EndEffDateTime',
        is_global=False,
        generate_sql=lambda config: f"SELECT * FROM `{config['full_table_name']}` WHERE {config['begin_date_column']} >= {config['end_date_column']} LIMIT 100"
    ),

    'scd2_unique_begin_date': TestTemplate(
        test_id='scd2_unique_begin_date',
        name='Unique begin datetime per Primary Key',
        category='integrity',
        severity='HIGH',
        description='Ensure no primary key has multiple records starting at the same time',
        is_global=False,
        generate_sql=lambda config: (
            f"""
            SELECT {' , '.join(config['primary_keys'])}, {config['begin_date_column']}, COUNT(*)
            FROM `{config['full_table_name']}`
            GROUP BY {' , '.join(config['primary_keys'])}, {config['begin_date_column']}
            HAVING COUNT(*) > 1
            LIMIT 100
            """
        )
    ),

    'scd2_unique_end_date': TestTemplate(
        test_id='scd2_unique_end_date',
        name='Unique end datetime per Primary Key',
        category='integrity',
        severity='HIGH',
        description='Ensure no primary key has multiple records ending at the same time',
        is_global=False,
        generate_sql=lambda config: (
            f"""
            SELECT {' , '.join(config['primary_keys'])}, {config['end_date_column']}, COUNT(*)
            FROM `{config['full_table_name']}`
            GROUP BY {' , '.join(config['primary_keys'])}, {config['end_date_column']}
            HAVING COUNT(*) > 1
            LIMIT 100
            """
        )
    ),

    'scd2_continuity': TestTemplate(
        test_id='scd2_continuity',
        name='Continuous history (no gaps)',
        category='validity',
        severity='HIGH',
        description='Check for gaps or overlaps in SCD history',
        is_global=False,
        generate_sql=lambda config: (
            f"""
            WITH ordered_history AS (
                SELECT 
                    *,
                    LEAD({config['begin_date_column']}) OVER (PARTITION BY {' , '.join(config['primary_keys'])} ORDER BY {config['begin_date_column']}) as next_begin
                FROM `{config['full_table_name']}`
            )
            SELECT * FROM ordered_history
            WHERE next_begin IS NOT NULL 
            AND DATE_ADD({config['end_date_column']}, INTERVAL 1 SECOND) <> next_begin
            LIMIT 100
            """
        )
    ),

    'scd2_no_record_after_current': TestTemplate(
        test_id='scd2_no_record_after_current',
        name='No record after current row',
        category='validity',
        severity='HIGH',
        description='Ensure no record exists after the high-watermark current row',
        is_global=False,
        generate_sql=lambda config: (
            f"""
            WITH history AS (
                SELECT 
                    *,
                    LEAD({config['begin_date_column']}) OVER (PARTITION BY {' , '.join(config['primary_keys'])} ORDER BY {config['begin_date_column']}) as next_begin
                FROM `{config['full_table_name']}`
            )
            SELECT * FROM history
            WHERE SAFE_CAST({config['active_flag_column']} AS STRING) IN ('true', 'TRUE', 'Y', '1')
            AND next_begin IS NOT NULL
            LIMIT 100
            """
        )
    ),

    # --- Structural Tests ---
    'surrogate_key_null': TestTemplate(
        test_id='surrogate_key_null',
        name='Surrogate key NOT NULL',
        category='completeness',
        severity='HIGH',
        description='Check surrogate key for NULL values',
        is_global=False,
        generate_sql=lambda config: f"SELECT * FROM `{config['full_table_name']}` WHERE {config['surrogate_key']} IS NULL LIMIT 100" if config.get('surrogate_key') else None
    ),

    'surrogate_key_unique': TestTemplate(
        test_id='surrogate_key_unique',
        name='Surrogate key uniqueness',
        category='integrity',
        severity='HIGH',
        description='Ensure surrogate key uniqueness',
        is_global=False,
        generate_sql=lambda config: (
            f"""
            SELECT {config['surrogate_key']}, COUNT(*) as duplicate_count
            FROM `{config['full_table_name']}`
            GROUP BY {config['surrogate_key']}
            HAVING COUNT(*) > 1
            LIMIT 100
            """ if config.get('surrogate_key') else None
        )
    )
}


def get_enabled_tests(enabled_test_ids: Optional[List[str]] = None) -> List[TestTemplate]:
    """
    Get enabled test templates.
    
    Args:
        enabled_test_ids: List of test IDs to enable. If None, returns all global tests.
        
    Returns:
        List of enabled test templates
    """
    if not enabled_test_ids:
        # Return all global tests by default
        return [test for test in PREDEFINED_TESTS.values() if test.is_global]
    
    return [
        PREDEFINED_TESTS[test_id]
        for test_id in enabled_test_ids
        if test_id in PREDEFINED_TESTS
    ]
